Doc: `Headers` documentation (#817)
zio/zio-http: docs/website/docs/v1.x/dsl/headers/index.md

# Headers

ZIO HTTP provides support for all HTTP (as defined in [RFC2616](https://datatracker.ietf.org/doc/html/rfc2616) ) headers and custom headers are supported too.

## Server-side

### Attaching Headers to `Response`
On the server-side, `ZIO-HTTP` is adding a collection of pre-defined headers to any response, according to the HTTP specification, but on top of them, users of the library may add other headers, including custom headers.

There are multiple ways to attach headers to a response:
- Using `addHeaders` helper on response.
```scala
val res = Response.ok.addHeader("X-MY-HEADER", "test")
```
- Through `Response` constructors.
```scala
val res = Response(
       status = Status.OK,
       // Setting response header 
       headers = Headers.contentLength(0L),
       data = HttpData.empty
```
- Using `Middlewares`.
```scala
val app = Http.ok @@ Middleware.addHeader("X-MY-HEADER", "test")
```
<details>

> <summary>Example below shows how the Headers could be added to a response by using `Response` constructors and how a custom header is added to `Response` through `addHeader` API call: </summary>
> <p>

Reviewer: ```suggestion
<summary>Detailed Examples</summary>
<p>
Example below shows how the Headers could be added to a response by using `Response` constructors and how a custom header is added to `Response` through `addHeader`: 
```

  val program = for {
    // Pass headers to request
    res  <- Client.request(url, headers)
    // List all response headers
    _    <- console.putStrLn(res.headers.toList.mkString("\n"))
    data <-
      // Check if response contains a specified header with a specified value.
      if (res.hasHeader(HeaderNames.contentType, HeaderValues.applicationJson))
        res.getBodyAsString
      else
        res.getBodyAsString
    _    <- console.putStrLn { data }
  } yield ()

  override def run(args: List[String]): URIO[zio.ZEnv, ExitCode] = program.exitCode.provideCustomLayer(env)

}
```
</p>
</details>

### Reading headers from `Response`

```scala
val responseHeaders: Task[Headers] =  Client.request(url).map(_.headers)
```

## Headers DSL

Headers DSL provides plenty of powerful operators that can be used to add, remove and modify headers. Same Headers API could be used on both client, server, and middleware.

`zhttp.http.Headers`      - represents an immutable collection of headers i.e. essentially a `Chunk[(String, String)]`.

`zhttp.http.HeaderNames`  - commonly used header names.

`zhttp.http.HeaderValues` - commonly used header values

> - Constructors - provides predefined constructors for most frequent used headers.

Reviewer: ```suggestion
- Constructors -  Provides a list of helpful methods that can create `Headers`.
```

```scala
import zhttp.http._

// create a simple Accept header:
val acceptHeader: Headers = Headers.accept(HeaderValues.applicationJson)

// create a basic authentication header:
val basicAuthHeader: Headers = Headers.basicAuthorizationHeader("username", "password")
```

> - Getters - provides predefined getters for most frequent used headers.

Reviewer: ```suggestion
- Getters - Provides a list of operators that parse and extract data from the headers.
```

```scala
import zhttp.http._

// retrieving the value of Accept header value:
val acceptHeader: Headers = Headers.accept(HeaderValues.applicationJson)
val acceptHeaderValue: Option[CharSequence] = acceptHeader.getAccept

// retrieving a bearer token from Authorization header:
val authorizationHeader: Headers                   = Headers.authorization("Bearer test")
val authorizationHeaderValue: Option[String]       = authorizationHeader.getBearerToken
```

> - Modifiers - provides `builder-pattern` oriented operations for most frequent used headers.

Reviewer: ```suggestion
- Modifiers - Provides a list of operators that modify the current Headers. Once modified, a new instance of the same type is returned.
```

```scala
import zhttp.http._

// add Accept header:
val headers = Headers.empty
val updatedHeadersList: Headers = headers.addHeaders(Headers.accept(HeaderValues.applicationJson))

// or if you prefer the builder pattern:

// add Host header:
val moreHeaders: Headers        = headers.withHost("zio-http.dream11.com")

```

> - Checks - provides operators that check if the Headers meet the given constraints.

Reviewer: ```suggestion
- Checks - Provides a list of operators that checks if the Headers meet the give constraints.
```

            )
          else { 
            // Adding a custom header to Response
            Response(status = Status.ACCEPTED, data = HttpData.fromChunk(message)).addHeader("X-MY-HEADER", "test")
          }
      }
    }
    
    ```

- The following example shows how Headers could be added to `Response` in the `Middleware` implementation:

    ```scala
    
      /**
       * Creates an authentication middleware that only allows authenticated requests to be passed on to the app.
       */
      final def customAuth(
        verify: Headers => Boolean,
        responseHeaders: Headers = Headers.empty,
      ): HttpMiddleware[Any, Nothing] =
        Middleware.ifThenElse[Request](req => verify(req.getHeaders))(
          _ => Middleware.identity,
          _ => Middleware.fromHttp(Http.status(Status.FORBIDDEN).addHeaders(responseHeaders)),
        )
    
    ```

- More examples:
  - [BasicAuth](https://github.com/dream11/zio-http/blob/main/example/src/main/scala/BasicAuth.scala)
  - [Authentication](https://github.com/dream11/zio-http/blob/main/example/src/main/scala/Authentication.scala)
</p>
</details>

> ## Client-Side

Reviewer: ```suggestion
## Client-side
```

    
      val program = for {
        // Pass headers to request
        res  <- Client.request(url, headers)
        // List all response headers
        _    <- console.putStrLn(res.headers.toList.mkString("\n"))
        data <-
          // Check if response contains a specified header with a specified value.
          if (res.hasHeader(HeaderNames.contentType, HeaderValues.applicationJson))
            res.getBodyAsString
          else
            res.getBodyAsString
        _    <- console.putStrLn { data }
      } yield ()
    
      override def run(args: List[String]): URIO[zio.ZEnv, ExitCode] = program.exitCode.provideCustomLayer(env)
    
    }
    ```
</p>
</details>

## Headers DSL

Headers DSL provides plenty of powerful operators that can be used to add, remove, modify and verify headers. Headers APIs could be used on client, server, and middleware.

`zhttp.http.Headers`      - represents an immutable collection of headers i.e. essentially a `Chunk[(String, String)]`.

`zhttp.http.HeaderNames`  - commonly used header names.

`zhttp.http.HeaderValues` - commonly used header values

> 

Reviewer: ```suggestion

`Headers` have following type of helpers
```
